### PR TITLE
[Merged by Bors] - feat(Data/List/Sort): `List.Sorted.cons`

### PR DESCRIPTION
**Pull request by vihdzp**
If `a :: l` is sorted under a transitive relation `r` and `r b a`, then `b :: a :: l` is sorted too.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Sort.lean
Comment (eric-wieser) [inline Mathlib/Data/List/Sort.lean:82]: Arguably this should be called `sorted_cons` to match `pairwise_cons`

Comment (Ruben-VandeVelde) [reply Mathlib/Data/List/Sort.lean:82]: I think this is more in line with lean 4, though

File: Mathlib/Data/List/Sort.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Sort.lean:76]: Please use `mem_cons` before `rintro`

File: Mathlib/Data/List/Sort.lean
Comment (eric-wieser) [inline Mathlib/Data/List/Sort.lean:80]:
```suggestion
nonrec theorem Sorted.cons {r : α → α → Prop} [IsTrans α r] {l : List α} {a b : α}
    (h : r b a) (ha : Sorted r (a :: l)) : Sorted r (b :: a :: l) :=
  ha.cons <| forall_mem_cons.2 ⟨h, fun _ hx => _root_.trans h <| rel_of_sorted_cons ha _ hx⟩
```

File: Mathlib/Data/List/Sort.lean
Comment (eric-wieser) [reply Mathlib/Data/List/Sort.lean:82]: Then we should rename `pairwise_cons` too

File: Mathlib/Data/List/Sort.lean
Comment (vihdzp) [reply Mathlib/Data/List/Sort.lean]: Out of curiosity, what does the `nonrec` keyword do?

File: Mathlib/Data/List/Sort.lean
Comment (eric-wieser) [inline Mathlib/Data/List/Sort.lean:75]: Some name tweaks:
```suggestion
nonrec theorem Sorted.cons {r : α → α → Prop} [IsTrans α r] {l : List α} {a b : α}
    (hab : r a b) (h : Sorted r (b :: l)) : Sorted r (a :: b :: l) :=
  h.cons <| forall_mem_cons.2 ⟨hab, fun _ hx => _root_.trans hab <| rel_of_sorted_cons h _ hx⟩
```

File: Mathlib/Data/List/Sort.lean
Comment (eric-wieser) [inline Mathlib/Data/List/Sort.lean:78]: This statement is backwards.

File: Mathlib/Data/List/Sort.lean
Comment (eric-wieser) [reply Mathlib/Data/List/Sort.lean]: It makes `.cons`  resolve to `Pairwise.cons` instead of the recursive `Sorted.cons`

File: Mathlib/Data/List/Sort.lean
Comment (mattrobball) [reply Mathlib/Data/List/Sort.lean:77]: Do you want to do that additional rename here? I don't.

Comment (vihdzp) [reply Mathlib/Data/List/Sort.lean:77]: Is that an uncontroversial rename? I can do it in a separate PR.

**Review by Ruben-VandeVelde — APPROVED** [review]
I think this can land as is, but defer to the maintainers

maintainer merge

File: Mathlib/Data/List/Sort.lean
Comment (eric-wieser) [reply Mathlib/Data/List/Sort.lean:77]: > I think this is more in line with lean 4, though

Revisiting, I don't know if I agree with this claim. I'll ping @YaelDillies for another opinion.

Comment (YaelDillies) [reply Mathlib/Data/List/Sort.lean:77]: I agree. This should be `sorted_cons`

**Comment by YaelDillies** [comment]
maintainer merge

File: Mathlib/Data/List/Sort.lean
Comment (fpvandoorn) [inline Mathlib/Data/List/Sort.lean:77]:
```suggestion
theorem sorted_cons_cons {r : α → α → Prop} [IsTrans α r] {l : List α} {a b : α} :
```
maybe?

**Comment by bryangingechen** [comment]
Thanks!
bors r+

**Pull request closed without merging**
